lib/livebook/hubs/team_client.ex

```diff
@@ -143,61 +163,159 @@ defmodule Livebook.Hubs.TeamClient do
+    Hubs.Broadcasts.file_system_created(file_system)
+
+    put_file_system(state, file_system)
+  end
+
+  defp handle_event(:file_system_created, file_system_created, state) do
+    handle_event(:file_system_created, build_file_system(state, file_system_created), state)
+  end
+
+  defp handle_event(:file_system_updated, %{external_id: _} = file_system, state) do
+    Hubs.Broadcasts.file_system_updated(file_system)
+
+    put_file_system(state, file_system)
+  end
+
+  defp handle_event(:file_system_updated, file_system_updated, state) do
+    handle_event(:file_system_updated, build_file_system(state, file_system_updated), state)
+  end
+
+  defp handle_event(:file_system_deleted, %{external_id: _} = file_system, state) do
+    Hubs.Broadcasts.file_system_deleted(file_system)
+
+    remove_file_system(state, file_system)
+  end
+
+  defp handle_event(:file_system_deleted, %{id: id}, state) do
+    if file_system = Enum.find(state.file_systems, &(&1.external_id == id)) do
+      handle_event(:file_system_deleted, file_system, state)
+    else
+      state
+    end
+  end
+
+  defp handle_event(:user_connected, user_connected, state) do
+    state
+    |> dispatch_secrets(user_connected)
+    |> dispatch_file_systems(user_connected)
+  end
+
+  defp dispatch_secrets(state, %{secrets: secrets}) do
+    decrypted_secrets = Enum.map(secrets, &build_secret(state, &1))
+
     created_secrets =
-      Enum.reject(secrets, fn secret ->
+      Enum.reject(decrypted_secrets, fn secret ->
         Enum.find(state.secrets, &(&1.name == secret.name and &1.value == secret.value))
       end)
 
     deleted_secrets =
       Enum.reject(state.secrets, fn secret ->
-        Enum.find(secrets, &(&1.name == secret.name))
+        Enum.find(decrypted_secrets, &(&1.name == secret.name))
       end)
 
     updated_secrets =
-      Enum.filter(secrets, fn secret ->
+      Enum.filter(decrypted_secrets, fn secret ->
         Enum.find(state.secrets, &(&1.name == secret.name and &1.value != secret.value))
       end)
 
-    secrets_by_topic = [
+    dispatch_events(state,
       secret_deleted: deleted_secrets,
       secret_created: created_secrets,
       secret_updated: updated_secrets
-    ]
+    )
+  end
+
+  defp dispatch_file_systems(state, %{file_systems: file_systems}) do
+    decrypted_file_systems = Enum.map(file_systems, &build_file_system(state, &1))
+
+    created_file_systems =
+      Enum.reject(decrypted_file_systems, fn file_system ->
+        Enum.find(state.file_systems, &(&1.external_id == file_system.external_id))
+      end)
+
+    deleted_file_systems =
+      Enum.reject(state.file_systems, fn file_system ->
+        Enum.find(decrypted_file_systems, &(&1.external_id == file_system.external_id))
+      end)
+
+    updated_file_systems =
+      Enum.filter(decrypted_file_systems, fn file_system ->
+        Enum.find(state.file_systems, &(&1.external_id == file_system.external_id))
+      end)
```

Comment:
We can encapsulate this on a function called:

    {created, deleted, updated} =
      diff(old, new, &(&1.external_id == &2.external_id))

and use it on both dispatching functions. :)